### ISSUE TITLE
AppDispatcher: Catastrophic failure (Exception from HRESULT: 0x8000FFFF (E_UNEXPECTED))

### ISSUE THREAD

**Issue by stevehansen** [RickStrahl/MarkdownMonster#999]
Seems I'm hitting this bug: https://github.com/MicrosoftEdge/WebView2Feedback/issues/2234

```
18/01/2023 21:51:36 - AppDispatcher: Catastrophic failure (Exception from HRESULT: 0x8000FFFF (E_UNEXPECTED))
Catastrophic failure (Exception from HRESULT: 0x8000FFFF (E_UNEXPECTED))
Markdown Monster v2.7.13.5
10.0.22000.1.amd64fre.co_release.210604-1628 - en-GB - NET 4.8 - 64 bit
en-BE - en-GB    
---
mscorlib
   at System.Runtime.InteropServices.Marshal.ThrowExceptionForHRInternal(Int32 errorCode, IntPtr errorInfo)
   at Microsoft.Web.WebView2.Core.CoreWebView2Environment.<CreateCoreWebView2ControllerAsync>d__84.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Web.WebView2.Wpf.WebView2.<>c__DisplayClass33_0.<<EnsureCoreWebView2Async>g__Init|0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MarkdownMonster.Windows.PreviewBrowser.WebViewPreviewHandler.<InitializeAsync>d__27.MoveNext() in D:\projects\MarkdownMonsterCode\MarkdownMonster\Windows\PreviewBrowser\WebViewPreviewHandler.cs:line 76
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)
System.Runtime.InteropServices.COMException
```

**Comment by RickStrahl** [comment]
Thanks  - I'm aware of these - this is the most common error I see however, I have no idea where it is coming from. It's a WebView crash, with no references back into the MM code base. It's the most common error I see in my logs, but I have yet to see it myself.

Do you have any context of what was going on when it happened?

I see you're on the latest 2.17.x release which uses the new async integration in MM, which I was hoping would minimize these failures 😦, but looks like that is still unrelated. I have open issues with the WebView team on this but nothing back from them either in a couple of years.

That said, these failures appear to be relatively rare. out of  ~1000 daily users I see 1 or 2 of these day perhaps.

**Comment by RickStrahl** [comment]
I'm going to run an experiment with different async instantiation methods. I've uploaded `2.7.13.6` with an update that uses a different way of calling the WebView initialization method that appears to be the source of the error. I missed that your error actually has the `InitializeAsync()` call sequence in it. Most of the others I see don't have that but the rest of the error trace looks identical.

Looking through logs I see two users with the same issue today - repeatedly it looks like - with multiple crashes spaced out over an hour or so. If one of those was you it would be helpful to find out if there was a certain operation that triggered it (or startup or?)...

Thanks,

**Comment by stevehansen** [comment]
could it be something recent, my VM now also throws a similar exception from the Microsoft Powertoys app when it tries to preview the .md files in explorer

```
System.Runtime.InteropServices.COMException (0x8000FFFF): Catastrophic failure (0x8000FFFF (E_UNEXPECTED))
   at System.Runtime.InteropServices.Marshal.ThrowExceptionForHR(Int32 errorCode)
   at Microsoft.Web.WebView2.Core.CoreWebView2Environment.CreateCoreWebView2ControllerAsync(IntPtr ParentWindow)
   at Microsoft.Web.WebView2.WinForms.WebView2.InitCoreWebView2Async(CoreWebView2Environment environment, CoreWebView2ControllerOptions controllerOptions)
   at Microsoft.PowerToys.PreviewHandler.Markdown.MarkdownPreviewHandlerControl.<>c__DisplayClass18_2`1.<<DoPreview>b__3>d.MoveNext()
--- End of stack trace from previous location ---
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__128_0(Object state)
```

The repo has a few issues logged in the last few days: https://github.com/microsoft/PowerToys/issues?q=is%3Aissue+is%3Aopen+E_UNEXPECTED

**Comment by RickStrahl** [comment]
If this is in a VM I wonder if it might have something to do with the GPU.

Can you try setting this setting in your Settings:

![image](https://user-images.githubusercontent.com/1374013/213537037-34f091f2-39bf-422f-a3b9-b6b9ce5bcf15.png)

Also as your original issue reference I am explicitly configuring DPI management, rather than using the default manifest processing - so overriding the `SetThreadDpiAwarenessContext()` (or the WPF equivalent anyway). I'll take a look - that was specifically done for older versions pre the 4.72 DPI improvements in the framework so that logic (and extra code) can probably be removed now. 

It's hard to test though.

What VM software are  you using? I test MM on a VM running on one of my servers on Vultr (I think they use some version of VMWare) and haven't seen any problems.

**Comment by RickStrahl** [comment]
So I took a look in MM in relation to the DPI settings, and I had already removed my custom DPI settings code a while back (probably once we switched to the 4.72 requirements). MM now runs all default DPI settings passed down from the OS into the app (which is DPI_AWARE).

Took a look at the crashes in PowerToys and the sequence is **exactly the same** on all of those as they are in MM. They always end up in:

```text
   at System.Runtime.InteropServices.Marshal.ThrowExceptionForHR(Int32 errorCode)
   at Microsoft.Web.WebView2.Core.CoreWebView2Environment.CreateCoreWebView2ControllerAsync(IntPtr ParentWindow)
   at Microsoft.Web.WebView2.WinForms.WebView2.InitCoreWebView2Async(CoreWebView2Environment environment, CoreWebView2ControllerOptions controllerOptions)
```

which is triggered (in MM at least) by the WebView Environment Creation (my error points at `EnsureCoreWebView2Async()`):

```cs
var env = await CoreWebView2Environment.CreateAsync(
    userDataFolder: wsApp.Constants.WebViewEnvironmentFolderName
);
await WebBrowser.EnsureCoreWebView2Async(env);
```

FWIW, this isn't new. I've been seeing this particular error in my logs for years - low volume but still there. It's not just for VM - I've had this occur on me once or twice but didn't connect it at the time until I looked at the logs later. It's a real bear to pin this one down for further investigation.

As said I did make a change in the way the initiallization code is called, which tweaks the timing of how the control is instantiated. This might have an effect. Checked my logs again today and I don't see any failures on the new version (which isn't saying much yet since it usually takes a couple of days for decent user uptake).

Thanks for your help - this is definitely helpful, if frustrating :smile:

**Comment by stevehansen** [comment]
Sorry for the confusion, on my VM I have no problems with MM. It was only the Powertoys that threw the similar expection.

It seems that I got stuck a bit on my normal machine, and it looks like they also have some recent issues with the latest version: https://github.com/MicrosoftEdge/WebView2Feedback/issues/3137
I can remember seeing some errors during the setup, seems like they leave the EdgeWebView folder in a bad state when you the web view is currently in use (in my case Outlook), and trying to repair it only made it worse.

Finally got to repairing the WebView2 without any edge code running, it repaired, and now MM is working correctly. Which I like but I can't reproduce the error anymore.

**Comment by RickStrahl** [comment]
MM doesn't use a shared WebView Environment folder - it uses a private one (by default in `%appdata%\Markdown Monster\`) so it shouldn't interfere with another application's WebView environment.

I thought you mentioned that MM was crashing repeatedly on startup but I guess you meant PowerToys?

**Comment by RickStrahl** [comment]
I'm cautiously optimistic that the recent fixes in 2.7.13.6+ and now 2.7.14 have perhaps mitigated the problems you've reported here.

Since making the async initialization changes I have not seen any new errors related to the WebView control. It's a little early to be sure, but given that it's been 2 days and I've not seen any errors on these newer versions is at least promising.

FWIW, the fixes involved the following:

* Switching the Previewer to use `Task InitializeAsync()` from `async void InitializeAsync()` 
* Calling the constructor and actually continuing on the task result

**Comment by RickStrahl** [comment]
So we're now into this for a couple of weeks and I'm not seeing **any** WebView related failures in my log for the new versions. 

Going to close this. If you end up running into these issues again, please re-open or create a new issue - we do want to know if there's still a latent issue here.
